File: clubseis-ui/src/backend/contabilidadService.js
import {appFetch, config} from './appFetch';

export const findAllConceptos = (onSuccess) =>
    appFetch('/contabilidad/conceptos', config('GET'), onSuccess);

export const createConcepto = (concepto, onSuccess, onErrors) =>
    appFetch('/contabilidad/conceptos', config('POST', concepto), onSuccess, onErrors);

export const updateConcepto = (concepto, onSuccess, onErrors) =>
    appFetch(`/contabilidad/conceptos/${concepto.id}`, config('PUT', concepto), onSuccess, onErrors);

export const findAllCategorias = (onSuccess) =>
    appFetch('/contabilidad/categorias', config('GET'), onSuccess);

export const createCategoria = (categoria, onSuccess, onErrors) =>
    appFetch('/contabilidad/categorias', config('POST', categoria), onSuccess, onErrors);

export const updateCategoria = (categoria, onSuccess, onErrors) =>
    appFetch(`/contabilidad/categorias/${categoria.id}`, config('PUT', categoria), onSuccess, onErrors);

export const findAllCuentas = (onSuccess) =>
    appFetch('/contabilidad/cuentas', config('GET'), onSuccess);

export const createCuenta = (cuenta, onSuccess, onErrors) =>
    appFetch('/contabilidad/cuentas', config('POST', cuenta), onSuccess, onErrors);

export const updateCuenta = (cuenta, onSuccess, onErrors) =>
    appFetch(`/contabilidad/cuentas/${cuenta.id}`, config('PUT', cuenta), onSuccess, onErrors);

export const findAllRazonSocial = (onSuccess) =>
    appFetch('/contabilidad/razon-social', config('GET'), onSuccess);

export const createRazonSocial = (razonSocial, onSuccess, onErrors) =>
    appFetch('/contabilidad/razon-social', config('POST', razonSocial), onSuccess, onErrors);

export const updateRazonSocial = (razonSocial, onSuccess, onErrors) =>
    appFetch(`/contabilidad/razon-social/${razonSocial.id}`, config('PUT', razonSocial), onSuccess, onErrors);

export const findMovimientos = ({razonSocialId, fecha, conceptoId, categoriaId, cuentaId, page, size, tipo},
                                onSuccess) => {

    let path = `/contabilidad/movimientos?page=${page}`;

    path += (tipo != null) ? `&tipo=${tipo}` : "";
    path += razonSocialId ? `&razonSocialId=${razonSocialId}` : "";
    path += fecha ? `&fecha=${fecha}` : "";
    path += conceptoId ? `&conceptoId=${conceptoId}` : "";
    path += categoriaId ? `&categoriaId=${categoriaId}` : "";
    path += cuentaId ? `&cuentaId=${cuentaId}` : "";
    path += size ? `&size=${size}` : "";

    return appFetch(path, config('GET'), onSuccess);

}

export const findMovimientoById = (id, onSuccess) =>
    appFetch(`/contabilidad/movimientos/${id}`, config('GET'), onSuccess);

export const createMovimiento = (movimiento, onSuccess, onErrors) =>
    appFetch('/contabilidad/movimientos', config('POST', movimiento), onSuccess, onErrors);

export const updateMovimiento = (movimiento, onSuccess, onErrors) =>
    appFetch(`/contabilidad/movimientos/${movimiento.id}`, config('PUT', movimiento), onSuccess, onErrors);

export const deleteMovimiento = (id, onSuccess, onErrors) =>
    appFetch(`/contabilidad/movimientos/${id}`, config('DELETE'), onSuccess, onErrors);

export const getResumen = (fechaInicio, fechaFin, onSuccess) =>
    appFetch(`/contabilidad/summary?fechaInicio=${fechaInicio}&fechaFin=${fechaFin}`, config('GET'), onSuccess);

export const dowloadExcel = ({razonSocialId, fecha, conceptoId, categoriaId, cuentaId, tipo},
                             onSuccess, onErrors) => {

    let path = `/contabilidad/descargar-excel?page=0`;

    path += (tipo != null) ? `&tipo=${tipo}` : "";
    path += razonSocialId ? `&razonSocialId=${razonSocialId}` : "";
    path += fecha ? `&fecha=${fecha}` : "";
    path += conceptoId ? `&conceptoId=${conceptoId}` : "";
    path += categoriaId ? `&categoriaId=${categoriaId}` : "";
    path += cuentaId ? `&cuentaId=${cuentaId}` : "";

    return appFetch(path, config('GET'), onSuccess, onErrors);
}

export const uploadExcel = (data, onSuccess, onErrors) =>
    appFetch('/contabilidad/subir-excel', config('POST', data), onSuccess, onErrors);

export const getFacturasBlock = ({keyword, page, size}, onSuccess, onErrors) => {

    let path = `/contabilidad/facturas?page=${page}&size=${size}`;

    path += keyword ? `&keyword=${encodeURIComponent(keyword)}` : "";

    return appFetch(path, config('GET'), onSuccess, onErrors);
}

export const getFacturaFile = (id, onSuccess, onErrors) =>
    appFetch(`/contabilidad/facturas/${id}`, config('GET'), onSuccess, onErrors);

File: backend/src/main/java/es/minstrel/app/model/services/ContabilidadService.java
package es.minstrel.app.model.services;

import es.minstrel.app.model.entities.*;
import es.minstrel.app.model.exceptions.DuplicateInstanceException;
import es.minstrel.app.model.exceptions.InstanceNotFoundException;
import es.minstrel.app.model.exceptions.UninitializedParameterException;
import es.minstrel.app.model.exceptions.UnsupportedFileTypeException;
import es.minstrel.app.model.services.utils.Block;
import es.minstrel.app.model.services.utils.FacturaItem;
import es.minstrel.app.model.services.utils.FileType;
import es.minstrel.app.model.services.utils.SummaryConta;

import java.io.IOException;
import java.io.InputStream;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Locale;

public interface ContabilidadService {

    List<Categoria> getAllCategorias();

    void createCategoria(String name)
            throws DuplicateInstanceException;

    void updateCategoria(Long id, String name)
            throws DuplicateInstanceException, InstanceNotFoundException;

    List<Concepto> getAllConcepto();

    void createConcepto(String name)
            throws DuplicateInstanceException;

    void updateConcepto(Long id, String name)
            throws DuplicateInstanceException, InstanceNotFoundException;

    List<RazonSocial> getAllRazonSocial();

    List<RazonSocial> getAllRazonSocial(String keywords);

    void createRazonSocial(String denominacion, String cifnif)
            throws DuplicateInstanceException;

    void updateRazonSocial(Long id, String denominacion, String cifnif)
            throws DuplicateInstanceException, InstanceNotFoundException;

    List<Cuenta> getAllCuentas();

    void createCuentas(String name)
            throws DuplicateInstanceException;

    void updateCuentas(Long id, String name)
            throws DuplicateInstanceException, InstanceNotFoundException;

    Block<Movimiento> getMovimientos(LocalDate fecha, Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId, Boolean tipo, int page, int size);

    Movimiento getMovimiento(Long id)
            throws InstanceNotFoundException;

    void createMovimiento(Movimiento movimiento, Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId,
                          Factura factura, String fileExtension, String fileContent)
            throws InstanceNotFoundException, UnsupportedFileTypeException, IOException;

    void updateMovimiento(Long id, LocalDate fecha, boolean esGasto, BigDecimal base0, BigDecimal base4, BigDecimal base10,
                          BigDecimal base21, Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId)
            throws InstanceNotFoundException;

    void deleteMovimiento(Long id)
            throws InstanceNotFoundException, IOException;

    SummaryConta getResumenBalance(LocalDate fechaInicio, LocalDate fechaFin);

    byte[] getExcel(LocalDate fecha, Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId, Boolean tipo)
            throws IOException;

    int uploadExcel(InputStream file) throws IOException;

    Block<Factura> getFacturasBlock(String keyword, int page, int size);

    FileType getFacturaFile(Long facturaId)
            throws InstanceNotFoundException, IOException;

    FileType createRecebi(String receptor, String receptorRol, String emisor, String cantidad, String concepto, Locale locale)
            throws IOException, UninitializedParameterException;

    FileType createRecebi(Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId, String receptor,
                          String receptorRol, String emisor, BigDecimal cantidad, String concepto, Locale locale)
            throws IOException, UninitializedParameterException, InstanceNotFoundException, UnsupportedFileTypeException;

    FileType createFactura(LocalDate fecha, String codigo, String receptor, List<FacturaItem> facturaItems, Locale locale)
            throws IOException, UninitializedParameterException;

    FileType createFactura(Long razonSocialId, Long conceptoId, Long categoriaId, Long cuentaId, LocalDate fecha,
                           String codigo, String receptor, List<FacturaItem> facturaItems, Locale locale)
            throws IOException, UninitializedParameterException, InstanceNotFoundException, UnsupportedFileTypeException;

}
